Class 1

In [52]:
ages = [19, 22, 21, 35, 28, 24, 31, 19, 27, 40]

#prints the number of people that participated in this research
len(ages)

10

In [53]:
#calculates the total ages
total_ages = sum(ages)
print(total_ages)

266


In [61]:
#calculate the mean of ages and round to 1d.p
average_age = total_ages / len(ages)

print(average_age.__round__(1))

26.6


In [65]:
#find minimum age
min(ages)

19

In [66]:
#find maximum age
max(ages)

40

In [71]:
#import statistics

import statistics as stat

In [72]:
#find median of ages

print(stat.median(ages))

25.5


Yes, the mean is higher(26.6) than median(25.5).

Because, the mean is sensitive to extreme values, however the median represents the middle point and ignores these extream values(outliers).

When the mean and median disagree, it's telling you that your data is not perfectly symmetrical and contains outliers.



Check-in:
1. Data Science is a multidisciplinary field that combines mathematics, statistics and programming to analyze, interpret and visualize large amounts of data.

2. Problem Definition
   Data Collection
   Data Preparation
   Exploration
   Analysis & Modelling
   Evaluation
   Deployment & Monitoring

3. Stages 2, 3 and 4 and this is because you have to source data that ticks yoour boxes, clean and enrich it(which takes alot of time) and then visualizing it(which can be stressful because you analyse based on what you want to know about the dataset).

4. One sector that interests me is Social Media.

Class 2

In [ ]:
#import pandas

import pandas as pd

In [ ]:
#pandas reads the csv file

df = pd.read_csv("50_startups.csv")

In [ ]:
#pandas display the file as a datframe and shows the first five data

df.head()

,R&D Spend,Administration,Marketing Spend,State,Profit
0,165349.20,136897.80,471784.10,New York,192261.83
1,162597.70,151377.59,443898.53,California,191792.06
2,153441.51,101145.55,407934.54,Florida,191050.39
3,144372.41,118671.85,383199.62,New York,182901.99
4,142107.34,91391.77,366168.42,Florida,166187.94


In [ ]:
# tells us the rows and columns

df.shape

(50, 5)

In [ ]:
# tells us about the columns, the counts and datatype. This is really helpful for quickly idntifying wrong data types.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     str    
 4   Profit           50 non-null     float64
dtypes: float64(4), str(1)
memory usage: 2.1 KB


R&D Spend = Continuous
Administration = Continuous
Marketing Spend = Continuous
State = Nominal
Profit = Continuous

In [ ]:
#checks null columns, no missing values.

df.isnull().sum()

R&D Spend          0
Administration     0
Marketing Spend    0
State              0
Profit             0
dtype: int64

In [ ]:
#takes state as series and marketing spend as dataframe.
s = df['State']
m = df[["Marketing Spend"]]

#prints type of state and marketing spend
print(type(s))  
print(type(m))

#prints their shapes
print(s.shape, m.shape)

<class 'pandas.Series'>
<class 'pandas.DataFrame'>
(50,) (50, 1)


In [ ]:
#gives us more info, the mean, median, count, min and max

df.describe()

,R&D Spend,Administration,Marketing Spend,Profit
count,50.000000,50.000000,50.000000,50.000000
mean,73721.615600,121344.639600,211025.097800,112012.639200
std,45902.256482,28017.802755,122290.310726,40306.180338
min,0.000000,51283.140000,0.000000,14681.400000
25%,39936.370000,103730.875000,129300.132500,90138.902500
50%,73051.080000,122699.795000,212716.240000,107978.190000
75%,101602.800000,144842.180000,299469.085000,139765.977500
max,165349.200000,182645.560000,471784.100000,192261.830000


None of the mean values look out of place when compared to their respective data spreads. If we look closely, the columns mean and median are in the same range.

In [ ]:
#export to json file
df.to_json("50_startups.csv.json", orient = "records", indent = 2)

#opens the file

pd.read_json("50_startups.csv.json")

,R&D Spend,Administration,Marketing Spend,State,Profit
0,165349.20,136897.80,471784.10,New York,192261.83
1,162597.70,151377.59,443898.53,California,191792.06
2,153441.51,101145.55,407934.54,Florida,191050.39
3,144372.41,118671.85,383199.62,New York,182901.99
4,142107.34,91391.77,366168.42,Florida,166187.94
5,131876.90,99814.71,362861.36,New York,156991.12
6,134615.46,147198.87,127716.82,California,156122.51
7,130298.13,145530.06,323876.68,Florida,155752.60
8,120542.52,148718.95,311613.29,New York,152211.77
9,123334.88,108679.17,304981.62,California,149759.96


Unlike the csv file, in the json file is stored in a nested, hierarchical list. Also the headers(state, profit..) appear repeatedly. The csv stores numeric and text data same, while the json file stores state in quotation marks and numeric without it.

1. School ID: It is just an identifier. It has no advantage or disadvantage on a student's result. It doesnt matter if a student registers early or late and has a ID card.

2. Home Address: Whether a student stays off-campus or on-campus doesnt affect your grades or change how well you read.

3. Outdoor Activities: A student going to parties or movie night doesn't necessarily mean he's not serious with his academics or he gets a lower grade.

Class 3

In [17]:
#import pandas

%pip install openpyxl
import pandas as pd


   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpyxl]
   -------------------- ------------------- 1/2 [openpy

In [18]:
df = pd.read_excel("50_startups_messy.xlsx")

In [ ]:
#prints the size of the dataset
df.shape

(50, 5)

In [20]:
#checks the number of missing values
df.isna().sum()

R&D Spend          0
Administration     0
Marketing Spend    1
State              0
Profit             0
dtype: int64

We can see that Marketing Spend has 1 null value.

In [23]:
#tells us the number of times a product or value appears
df.duplicated().sum()

np.int64(0)

No value is duplicated.

In [44]:
df.describe()

,R&D Spend,Administration,Marketing Spend,Profit
count,50.000000,50.000000,47.000000,50.000000
mean,73721.615600,121344.639600,207414.790851,112012.639200
std,45902.256482,28017.802755,123051.179029,40306.180338
min,0.000000,51283.140000,0.000000,14681.400000
25%,39936.370000,103730.875000,122932.510000,90138.902500
50%,73051.080000,122699.795000,210797.670000,107978.190000
75%,101602.800000,144842.180000,297889.465000,139765.977500
max,165349.200000,182645.560000,471784.100000,192261.830000


In [25]:
#standardising state
df["State"] = df["State"].str.strip().str.title()

In [39]:
#coercing marketing spend to numeric
df["Marketing Spend"] = pd.to_numeric(df["Marketing Spend"], errors="coerce")

In [35]:
#simple random
srs = df.sample(n=20,  random_state=7)

#stratified, 20% of each state
strat = df.groupby("State", group_keys=False ) \
      .apply (
          lambda g: g.sample(frac=0.2))

In [ ]:
#comapring mean of r&d and marketing spend, with their populations and samples
compare = pd.DataFrame({
    "population": df[["Marketing Spend", "R&D Spend"]].mean(),
    "sample": srs[["Marketing Spend", "R&D Spend"]].mean()
})
print(compare.round(3))

                 population      sample
Marketing Spend  207414.791  239540.931
R&D Spend         73721.616   76076.374


R&D Spend is much closer. The sample and population's mean has a difference of 2354.758 while for Marketing Spend, the difference is about 32126.14(which is very large).

Why?
Marketing Spend has large standard deviation(123051.179029) which is nearly 3 times than R&D Spend's standard deviation. When a data has massive range, a small random samole is likely to pull extreme values that swerve the sample mean from the true populaton mean. 

Check-in:
1. A population is everything or every single unit the question is about, while the sampling frame is like a subset of population, the list of people who match your criteria and want to use for maybe survey, interview and other things.

2. A situation where secondary data is a better choice is if you're working for a company that deals with large datasets, like for instance sales dataset. You can't possibly start gathering information from people all around the world to find out where large profit or sales is. It's much better to download your preferred dataset on kaggle, clean and enrich it to  your preference. You can even use PRISCA AI to clean, enrich, add or remove columns, visualize and deploy, can also create dashboards.

A situation where it's not better is if you just want to get information from a few set of people, probably a feedback, you'll need to source it primarily.

3. Stratified samples some units from every group while cluster samples every units from some groups.

4. Uniqueness - df.duplicated().sum(),
   Completeness - df.isna().sum(),
   Validity - score = "N/A"
   